### DISCUSSION TITLE
Debugger advice

### DISCUSSION THREAD

**Discussion by barticus** [gadget-inc/wds#19] [Q&A] [answered]
Hi!

I switched a project to using esbuild-dev over ts-node-dev to detect some esbuild specific issues earlier in the dev cycle. It's really helped with that so thanks so much for putting this together!

In vscode, with ts-node-dev we used the node debugger. With esbuild-dev, the debugger is attached but line numbers don't match and it goes to the compiled JS not to the source files.
I'm guessing this is because source maps are not made by default.
Do you have any advice on how I can enable source maps through esbuild-dev? Or other work arounds?

Thanks,

**Comment by airhorns** [comment]
Hi @barticus glad the tools is helpful! This is strange! What debugger are you using? We use the one built into vscode running processes via launch.json as well as via the jest extension to debug tests, and we get dropped into typescript code with correct line numbers. Which version of `esbuild-dev` are you on as well?

**Comment by barticus** [comment]
Hi @airhorns , thanks for getting back to me.
My launch.json is:

```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch API",
            "request": "launch",
            "runtimeExecutable": "yarn",
            "runtimeArgs": [
              "api",
              "dev"
            ],
            "type": "node",
        }
    ]
}
```

Notes:
1. "yarn api dev" starts esbuild in the api yarn workspace
2. I've tried both node and pwa-node as the "type"
3. I do get lots of "Could not read source map for ... ENOENT: no such file or directory, open ..." where ... is the entry file, then the dependency files from that
4. Currently using esbuild-dev 0.8.0, ~i'll upgrade and see how that goes~ i see thats the latest version
5. I do have `sourceMap`, `declaration` and `declarationMap` settings to true in my tsconfig
6. I do have an outDir set, but esbuild-dev doesnt seem to generate anything there anyway

Thanks,

**Reply by airhorns** [reply]
Ok, this is our launch.json that works with esbuild-dev:
```
    {
      "type": "node",
      "name": "api-server",
      "request": "launch",
      "args": ["--supervise", "src/server.ts"],
      "cwd": "${workspaceFolder}/packages/api",
      "console": "integratedTerminal",
      "internalConsoleOptions": "neverOpen",
      "disableOptimisticBPs": true,
      "outFiles": [],
      "skipFiles": ["<node_internals>/**/*.js"],
      "program": "${workspaceFolder}/node_modules/.bin/esbuild-dev"
    },
```

We use yarn workspaces so thats why the `cwd` is set to something else. Also, do you have an `esbuild-dev.js` config file by any chance? You can also try running with `ESBUILD_DEV_DEBUG=1` to get esbuild-dev to log a bunch more details about what its doing and where it is writing temporary data to so you can inspect it.

**Reply by barticus** [reply]
Thanks @airhorns , I'll have a play around and see if that also works for me. As mentioned, i'm also in yarn workspaces but am using the individual yarn workspace to run esbuild-dev, not the root (i see you've set the program to point to your root node_modules).
My `dev` command expands to `esbuild-dev -r tsconfig-paths/register -s true app.ts`, i needed the register paths so it wouldnt freak out about `paths` set in tsconfig.

RE esbuild-dev.js, no I don't have one, in fact i didn't know you could! Does this take similar structure to an input file to esbuild?

Thanks!

**Comment by barticus** [comment]
Hi @airhorns ,

I've tried your above configuration and am still getting the same issue.

I was wondering if you could provide an example for esbuild-dev.js?
Also, whether you use `source-map-support/register` in any of your files? I've attempted to add to no avail/effect.

Thanks,

**Comment by barticus** [comment, marked as answer]
Latest build (^0.9.1) seems to have resolved this issue.
